File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation

def recv_all(socket: socket.socket, size):
    bytes_received = 0
    buffer = bytearray()
    while bytes_received < size: 
        data = socket.recv(size - bytes_received)
        if not data:
            raise ConnectionError
        bytes_received += len(data)
        buffer.extend(data)
    return bytes(buffer)


def send_all(socket: socket.socket, bytes):
    bytes_sent = 0
    while bytes_sent < len(bytes): 
        bytes_written = socket.send(bytes[bytes_sent:])
        if bytes_written == 0:
            continue
        bytes_sent += bytes_written
    

    

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	bytes_sent := 0
	for bytes_sent < len(bytes) {
		bytes_written, err := socket.Write(bytes[bytes_sent:])
		if err != nil {
			return err
		}
		bytes_sent += bytes_written
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	bytes_received := 0
	buff := make([]byte, size)
	for bytes_received < size {
		bytes_read, err := socket.Read(buff[bytes_received:])
		bytes_received += bytes_read
		if err != nil {
			if bytes_received == size {
				break
			}
			return nil, err
		}
	}
	return buff, nil
}
